File: SaplingViewModel.java
package com.internshipproject;

import android.app.Application;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;

import java.util.List;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class SaplingViewModel extends AndroidViewModel {
    private final SaplingDao saplingDao;
    private final LiveData<List<Sapling>> allSaplings;
    private final ExecutorService executorService;

    public SaplingViewModel(@NonNull Application application) {
        super(application);
        AppDatabase db = AppDatabase.getInstance(application);
        saplingDao = db.saplingDao();
        allSaplings = saplingDao.getAllSaplings();
        executorService = Executors.newSingleThreadExecutor();
    }

    public LiveData<List<Sapling>> getAllSaplings() {
        return allSaplings;
    }

    public void insert(Sapling sapling) {
        executorService.execute(() -> saplingDao.insert(sapling));
    }

    public void update(Sapling sapling) {
        executorService.execute(() -> saplingDao.update(sapling));
    }
}

File: SaplingViewModel.kt
package com.internshipproject

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.LiveData
import androidx.lifecycle.viewModelScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch

class SaplingViewModel(application: Application) : AndroidViewModel(application) {
    private val saplingDao: SaplingDao = AppDatabase.getInstance(application).saplingDao()
    val allSaplings: LiveData<List<Sapling>> = saplingDao.getAllSaplings()

    fun insert(sapling: Sapling) = viewModelScope.launch(Dispatchers.IO) {
        saplingDao.insert(sapling)
    }

    fun update(sapling: Sapling) = viewModelScope.launch(Dispatchers.IO) {
        saplingDao.update(sapling)
    }
}
